python/lsst/obs/decam/rawFormatter.py

```diff
@@ -22,18 +22,113 @@
+                   50: 50, 47: 51, 48: 52, 52: 53, 53: 54, 54: 55, 55: 56, 57: 57, 58: 58, 59: 59,
+                   60: 60, 61: 61, 62: 62, 72: 63, 71: 64, 64: 65, 63: 66, 73: 67, 74: 68, 70: 69,
+                   69: 70
+                   }
+
+
 class DarkEnergyCameraRawFormatter(FitsRawFormatterBase):
-    translatorClass = DecamTranslator
+    translatorClass = astro_metadata_translator.DecamTranslator
     filterDefinitions = DarkEnergyCamera.filterDefinitions
 
     def getDetector(self, id):
         return DarkEnergyCamera().getCamera()[id]
+
+    def _scanHdus(self, filename, detectorId):
+        """Scan through a file for the HDU containing data from one detector.
+
+        Parameters
+        ----------
+        filename : `str`
+            The file to search through.
+        detectorId : `int`
+            The detector id to search for.
+
+        Returns
+        -------
+        fitsData: `lsst.afw.fits.Fits`
+            An read-only struct for accessing the required HDU.
+        metadata: `lsst.daf.base.PropertyList`
+            The metadata read from the header for that detector id.
+
+        Raises
+        ------
+        ValueError
+            Raised if detectorId is not found in any of the file HDUs
+        """
+        log = lsst.log.Log.getLogger("DarkEnergyCameraRawFormatter")
+        log.debug("Did not find detector=%s at expected HDU=%s in %s: scanning through all HDUs.",
+                  detectorId, detector_to_hdu[detectorId], filename)
+
+        fitsData = lsst.afw.fits.Fits(filename, 'r')
+        # NOTE: The primary header (HDU=0) does not contain detector data.
+        for i in range(1, fitsData.countHdus()):
+            fitsData.setHdu(i)
+            metadata = fitsData.readMetadata()
+            if metadata['CCDNUM'] == detectorId:
+                return fitsData, metadata
+        else:
+            raise ValueError(f"Did not find detectorId={detectorId} as CCDNUM in any HDU of {filename}.")
+
+    def _determineHDU(self, detectorId):
+        """Determine the correct HDU number for a given detector id.
+
+        Parameters
+        ----------
+        detectorId : `int`
+            The detector id to search for.
+
+        Returns
+        -------
+        fitsData : `lsst.afw.fits.Fits`
+            An read-only struct for accessing the required HDU.
+        metadata : `lsst.daf.base.PropertyList`
+            The metadata read from the header for that detector id.
+
+        Raises
+        ------
+        ValueError
+            Raised if detectorId is not found in any of the file HDUs
+        """
+        filename = self.fileDescriptor.location.path
+        try:
+            metadata = lsst.afw.image.readMetadata(filename, detector_to_hdu[detectorId])
+            if metadata['CCDNUM'] != detectorId:
+                # the detector->HDU mapping is different in this file: try scaning
+                return self._scanHdus(filename, detectorId)
```

Comment:
Good catch. I thought this one would be caught by the test with the original file in `testdata_decam`, but that goes into the `except FitsError` branch (the data is not in order, but there are also only 2 HDUs). I could try making another stripped file and randomizing the order of the HDUs to trigger this path. I don't think it would be easy to mock out `readMetadata` and `fits.Fits` in a useful manner.

Reply:
I think copy the stripped file and swap two of the CCDNUM values over. Then ask for the data from one of the swapped values.

Reply:
I've gone further: I made a copy of the stripped file and shuffled all the HDUs and now have a test that compares reading the metadata with the non-shuffled version.